File: JavaScript/StudyPlan/LeetCode75/Q735-AsteroidCollision.js
/**
 * @param {number[]} asteroids
 * @return {number[]}
 */
var asteroidCollision = function(asteroids) {
    const stars = [];

    for(let i = 0; i < asteroids.length; i++) {
        const a = asteroids[i];
        if((stars.length === 0 || stars[stars.length - 1] < 0) && a < 0)  {
            stars.push(a);
        } else if(a > 0) {
            stars.push(a);
        } else {
            const star = stars.pop();
            
            if(Math.abs(star) > Math.abs(a)) {
                stars.push(star);
            } else if(Math.abs(star) < Math.abs(a)) {
                i--;
            } else {
                continue;
            }
        }
    }
    return stars;
};

File: Python/Leetcode75/Medium/Stack/Q735-AsteroidCollision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        collides = []

        for asteroid in asteroids:
            while True:
                if len(collides) == 0 or asteroid > 0 or collides[-1] < 0:
                    collides.append(asteroid)
                    break
                elif collides[-1] > 0:
                    if collides[-1] < -asteroid:
                        collides.pop()
                        continue
                    elif collides[-1] == -asteroid:
                        collides.pop()
                    break
                else:
                    collides.append(asteroid)
                    break

        return collides